src/apps/lwaftr/conf_parser.lua

```diff
@@ -0,0 +1,292 @@
+   self.peek_char = self.read_char()
+   return chr
+end
+
+function Parser:peek() return self.peek_char end
+function Parser:is_eof() return not self:peek() end
+
+function Parser:check(expected)
+   if self:peek() == expected then
+      if expected then self:next() end
+      return true
+   end
+   return false
+end
+
+function Parser:consume(expected)
+   if not self:check(expected) then
+      local ch = self:peek()
+      if ch == nil then
+         self:error("while looking for '%s', got EOF", expected)
+      else
+         self:error("expected '%s', got '%s'", expected, ch)
+      end
+   end
+end
+
+function Parser:take_while(pattern)
+   local res = {}
+   while not self:is_eof() and self:peek():match(pattern) do
+      table.insert(res, self:next())
+   end
+   return table.concat(res)
+end
+
+function Parser:skip_whitespace()
+   self:take_while('%s')
+   -- Skip comments, which start with # and continue to the end of line.
+   while self:check('#') do
+      self:take_while('[^\n]')
+      self:take_while('%s')
+   end
+end
+
+function Parser:parse_uint(min, max)
+   local tok = self:take_while('%d')
+   if tok == '' then self:error('expected a number') end
+   if #tok > #(tostring(max)) then
+      self:error('numeric constant too long: %s', tok)
+   end
+   local uint = tonumber(tok)
+   if uint < min or uint > max then
+      self:error('numeric constant out of range: %d', uint)
+   end
+   return uint
+end
+
+function Parser:parse_psid_param() return self:parse_uint(0, 16) end
+function Parser:parse_ipv4_quad() return self:parse_uint(0, 255) end
+
+function Parser:parse_property_list(spec, bra, ket)
+   local res = {}
+   self:skip_whitespace()
+   if bra then
+      self:consume(bra)
+      self:skip_whitespace()
+   end
+   while not self:check(ket) do
+      local key = self:take_while('[%w_]')
+      if key == '' then
+         self:error("expected a key=value property or a closing '%s'", ket)
+      end
+      if res[key] then self:error('duplicate key: %s', key) end
+      if not spec.parse[key] then self:error('unexpected key: %s', key) end
+      self:skip_whitespace()
+      self:consume('=')
+      self:skip_whitespace()
+      local val = spec.parse[key](self)
+      res[key] = val
+
+      -- Key-value pairs are separated by newlines or commas, and
+      -- terminated by the ket.  A trailing comma is optional.
+      local line = self.line
+      self:skip_whitespace()
+      local has_comma = self:check(',')
+      if has_comma then self:skip_whitespace() end
+      if self:check(ket) then break end
+      if not has_comma and self.line == line then
+         self:error('properties should be separated by commas or newlines')
+      end
+   end
+   for k, default in pairs(spec.defaults) do
+      if res[k] == nil then res[k] = default(res) end
+   end
+   spec.validate(self, res)
+   return res
+end
+
+-- Returns a uint8_t[4].
+function Parser:parse_ipv4()
+   local addr, err = ipv4:pton(self:take_while('[0-9.]'))
```

Comment:
Can use wildcard '%d' to match digits, unless you'd like to do it this way (more explicit?)